File: Problems 601 to 650/Set Mismatch.py
# https://leetcode.com/problems/set-mismatch

# Runtime: 207 ms, faster than 89.89% of Python3 online submissions for Set Mismatch.
# Good runtime, the constant checking of a set is likely not faster than alternatives.

# Memory Usage: 15.8 MB, less than 55.38% of Python3 online submissions for Set Mismatch.
# Average memory usage, lots of variables are created, and the set is also contributing to this.


class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        # thanks to: https://dev.to/seanpgallivan/solution-set-mismatch-4ejh
        
        repeated = 0
        seen = set()
        N = len(nums)
        sumN = (N * (N+1)) // 2
        
        for n in nums:
            sumN -= n
            
            if n in seen:
                repeated = n
            seen.add(n)
            
            
        return [repeated, sumN + repeated]

File: Problems 601 to 650/Set Mismatch.java
class Solution {
    public int[] findErrorNums(int[] nums) {

        int repeated = 0;
        var seen = new HashSet<Integer>();
        int N = nums.length;
        int sumNums = (N * (N+1)) / 2;

        for (int num : nums) {
            sumNums -= num;

            if (seen.contains(num)) {
                repeated = num;
            }
            seen.add(num);
        }

        int[] result = {repeated, sumNums + repeated};
        return result;
    }
}
